### PR TITLE
bot initial structure 

### PR DESCRIPTION
**Pull request by gargatuma**
Created architecture of bot for further divison of tasks among interns

### CODE REVIEW

File: src/main/java/com/exadel/coolDesking/bot/config/botConfig.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import org.telegram.telegrambots.meta.TelegramBotsApi;
+import org.telegram.telegrambots.meta.api.methods.updates.SetWebhook;
+import org.telegram.telegrambots.meta.exceptions.TelegramApiException;
+import org.telegram.telegrambots.updatesreceivers.DefaultBotSession;
+import org.telegram.telegrambots.updatesreceivers.DefaultWebhook;
+
+@Configuration
+public class botConfig {
+    private final String WEBHOOK = System.getenv("WEBHOOK");
+    private final String URL = System.getenv("URL");
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/config/botConfig.java:16]: https://www.baeldung.com/spring-value-annotation

Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/config/botConfig.java:16]: use more meaningful names for the env vars like `TELEGRAM_WEBHOOK_URL` and `TELEGRAM_INTERNAL_URL`

File: src/main/java/com/exadel/coolDesking/bot/HotDeskingBot.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package com.exadel.coolDesking.bot;
+
+import org.telegram.telegrambots.meta.api.methods.BotApiMethod;
+import org.telegram.telegrambots.meta.api.methods.updates.SetWebhook;
+import org.telegram.telegrambots.meta.api.objects.Update;
+import org.telegram.telegrambots.starter.SpringWebhookBot;
+
+public class HotDeskingBot extends SpringWebhookBot {
+    private final String BOT_USERNAME = System.getenv("BOT_USERNAME");
+    private final String BOT_TOKEN = System.getenv("BOT_TOKEN");
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/HotDeskingBot.java:10]: move to constructor + https://www.baeldung.com/spring-value-annotation

File: src/main/java/com/exadel/coolDesking/bot/BotController.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import org.springframework.web.bind.annotation.RestController;
+import org.telegram.telegrambots.meta.api.methods.BotApiMethod;
+import org.telegram.telegrambots.meta.api.objects.Update;
+
+@RequiredArgsConstructor
+@RestController
+public class BotController {
+    private final HotDeskingBot hotDeskingBot;
+
+    @PostMapping("/")
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/BotController.java:15]: any chance to use another mapping? like `/api/telegram-bot`

File: src/main/java/com/exadel/coolDesking/bot/HotDeskingBot.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    @Override
+    public BotApiMethod<?> onWebhookUpdateReceived(Update update) {
+        return updateReceiver.handleUpdate(update);
+    }
+
+    @Override
+    public String getBotPath() {
+        return "exadel";
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/HotDeskingBot.java:35]: at least make it constant

File: src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+@Slf4j
+@Component
+public class UpdateReceiver {
+    private final UserService userService;
+    private final Map<UserState, Handler> handlers = new HashMap<>();
+
+    public UpdateReceiver(UserService service, List<Handler> handlers) {
+        userService = service;
+        handlers.forEach(handler ->
+                this.handlers.put(handler.getHandlerStateName(), handler));
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java:28]: try using `.stream().collect(toMap(...))`

File: src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+            telegramId = query.getFrom().getId().toString();
+            log.info("TelegramId: " + telegramId);
+            user = userService.getByTelegramId(telegramId);
+            log.info("UserState: " + user.getUserState());
+            handler = handlers.get(user.getUserState());
+            log.info("Handler: " + handler);
+            response= handler.handle(query, user);
+
+            userService.updateUserStatus(response.getUserState(), telegramId);
+            System.out.println(response.getMethod());
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java:49]: use logger

File: src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+    public BotApiMethod<?> handleUpdate(Update update) {
+
+        String telegramId;
+        User user;
+        Handler handler;
+        BotResponseDto response;
+        if (update.hasCallbackQuery()) {
+            CallbackQuery query = update.getCallbackQuery();
+            log.info("DATA: " + query.getData());
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java:39]: `log.info("DATA: {}", query.getData());`

File: src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+            telegramId = query.getFrom().getId().toString();
+            log.info("TelegramId: " + telegramId);
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java:41]: same as with data DATA

File: src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+            user = userService.getByTelegramId(telegramId);
+            log.info("UserState: " + user.getUserState());
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java:43]: same as with data DATA

File: src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+            handler = handlers.get(user.getUserState());
+            log.info("Handler: " + handler);
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java:45]: same as with data DATA

Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java:45]: are you sure any of this data is not sensitive and can be logged?

File: src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+                this.handlers.put(handler.getHandlerStateName(), handler));
+    }
+
+    public BotApiMethod<?> handleUpdate(Update update) {
+
+        String telegramId;
+        User user;
+        Handler handler;
+        BotResponseDto response;
+        if (update.hasCallbackQuery()) {
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java:37]: can be moved into a separate method

File: src/main/java/com/exadel/coolDesking/bot/config/botConfig.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        HotDeskingBot hotDeskingBot = new HotDeskingBot(setWebhookInstance, updateReceiver);
+        DefaultWebhook defaultWebhook = new DefaultWebhook();
+        defaultWebhook.setInternalUrl(URL);
+        try {
+            TelegramBotsApi telegramBotsApi = new TelegramBotsApi(DefaultBotSession.class);
+            telegramBotsApi.registerBot(hotDeskingBot, setWebhookInstance);
+            return hotDeskingBot;
+        } catch (TelegramApiException e) {
+            e.printStackTrace();
+            return null;
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/config/botConfig.java:39]: maybe you want to fail the application startup if the bot can't be loaded ? You need to rethrow the exception then.

File: src/main/java/com/exadel/coolDesking/bot/config/botConfig.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    ) {
+        HotDeskingBot hotDeskingBot = new HotDeskingBot(setWebhookInstance, updateReceiver);
+        DefaultWebhook defaultWebhook = new DefaultWebhook();
+        defaultWebhook.setInternalUrl(URL);
+        try {
+            TelegramBotsApi telegramBotsApi = new TelegramBotsApi(DefaultBotSession.class);
+            telegramBotsApi.registerBot(hotDeskingBot, setWebhookInstance);
+            return hotDeskingBot;
+        } catch (TelegramApiException e) {
+            e.printStackTrace();
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/config/botConfig.java:38]: log it properly

File: src/main/java/com/exadel/coolDesking/bot/dto/BotResponseDto.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+import lombok.Getter;
+import org.springframework.stereotype.Component;
+import org.telegram.telegrambots.meta.api.methods.BotApiMethod;
+import org.telegram.telegrambots.meta.api.methods.send.SendMessage;
+import org.telegram.telegrambots.meta.api.methods.updatingmessages.EditMessageText;
+import org.telegram.telegrambots.meta.api.objects.replykeyboard.InlineKeyboardMarkup;
+
+@Getter
+@Component
+public class BotResponseDto {
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/dto/BotResponseDto.java:13]: stateful components should not be singletons. why is it a component anyway?

File: src/main/java/com/exadel/coolDesking/bot/exception/BotExceptionAdvice.java [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+@Slf4j
+@RestControllerAdvice
+public class BotExceptionAdvice {
+    @ExceptionHandler(BotNotFoundException.class)
+    @ResponseStatus(HttpStatus.OK)
+    public BotApiMethod<?> handleBotNotFound(BotNotFoundException exception) {
+        log.warn(exception.getMessage() +
+                "; Telegram Id: " + exception.getTelegramId() +
+                "; Class: " + exception.getType()
+                );
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/exception/BotExceptionAdvice.java:21]: `log.warn("{}; Telegram Id: {}; Class: {}",exception.getMessage(), exception.getTelegramId(), exception.getType());`

File: src/main/java/com/exadel/coolDesking/bot/exception/BotExceptionAdvice.java [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+                .build();
+    }
+
+    @ExceptionHandler(BotBadRequestException.class)
+    @ResponseStatus(HttpStatus.OK)
+    public BotApiMethod<?> botBadRequestHandler(BotBadRequestException exception) {
+        log.warn(exception.getMessage() +
+                "; Telegram Id: " + exception.getTelegramId() +
+                "; Class: " + exception.getType()
+        );
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/exception/BotExceptionAdvice.java:34]: same as in handleBotNotFound

File: src/main/java/com/exadel/coolDesking/bot/handler/city/CityHandler.java [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+    private final CityHandlerService service;
+
+    @Override
+    public BotResponseDto handle(CallbackQuery query, User user) {
+        String cityName = query.getData();
+        Integer messageId = query.getMessage().getMessageId();
+        String chatId = query.getMessage().getChatId().toString();
+
+        if (service.isBack(cityName))
+            return service.backToMainMenu(messageId, chatId);
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/handler/city/CityHandler.java:23]: parenthesis

File: src/main/java/com/exadel/coolDesking/bot/handler/country/CountryHandler.java [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+public class CountryHandler implements Handler {
+    private final CountryHandlerService service;
+
+    @Override
+    public BotResponseDto handle(CallbackQuery query, User user) {
+        String countryName = query.getData();
+        Integer messageId = query.getMessage().getMessageId();
+        String chatId = query.getMessage().getChatId().toString();
+        if (service.isBack(countryName))
+            return service.backToMainMenu(messageId, chatId);
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/handler/country/CountryHandler.java:23]: parenthesis

File: src/main/java/com/exadel/coolDesking/bot/handler/country/CountryHandlerService.java [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+    private final OfficeService officeService;
+    private final BotResponseDto response;
+
+    public BotResponseDto showListOfCities(Integer messageId, String countryName, String telegramId, String chatId) {
+        List<String> allCities = officeService.getCitiesByCountry(countryName);
+        if (allCities.isEmpty())
+            throw new BotNotFoundException(
+                    MessageService.COUNTRY_NOT_FOUND + countryName,
+                    CityHandler.class,
+                    telegramId);
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/handler/country/CountryHandlerService.java:27]: parenthesis

File: src/main/java/com/exadel/coolDesking/bot/handler/city/CityHandlerService.java [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+    private final BotResponseDto response;
+    private final OfficeService officeService;
+
+    public BotResponseDto showListOfOffices(Integer messageId, String cityName, String telegramId, String chatId) {
+        List<String> officeNames = officeService.getAllNames(cityName);
+        if (officeNames.isEmpty())
+            throw new BotNotFoundException(
+                    MessageService.OFFICE_NOT_FOUND + cityName,
+                    CityHandler.class,
+                    telegramId);
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/handler/city/CityHandlerService.java:26]: parenthesis

File: src/main/java/com/exadel/coolDesking/bot/handler/main/MainMenuHandler.java [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+    private final MainMenuHandlerService service;
+
+    @Override
+    public BotResponseDto handle(CallbackQuery query, User user) {
+        String text = query.getData();
+        Integer messageId = query.getMessage().getMessageId();
+        String chatId = query.getMessage().getChatId().toString();
+
+        if (service.isBook(text))
+            return service.showListOfCountries(messageId, chatId);
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/handler/main/MainMenuHandler.java:26]: parenthesis

File: src/main/java/com/exadel/coolDesking/bot/util/MarkupService.java [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+    static InlineKeyboardMarkup listToKeyboard(List<String> data) {
+        List<List<InlineKeyboardButton>> keyboardButtons = new ArrayList<>();
+
+        int i = 0;
+        for (; i <= data.size() - 2; i+=2) {
+            keyboardButtons.add(twoButtons(data.get(i), data.get(i+1)));
+        }
+
+        if (i == data.size() - 1)
+            keyboardButtons.add((List.of(oneButton(data.get(i), data.get(i)))));
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/util/MarkupService.java:27]: parenthesis

**Review by ohalushchak-exadel — CHANGES_REQUESTED** [review]

File: src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java (hunk 10 above)
Comment (gargatuma) [reply src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java:45]: I will remove them. I used them just for debugging.

File: src/main/java/com/exadel/coolDesking/bot/HotDeskingBot.java [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+import org.telegram.telegrambots.meta.api.objects.Update;
+import org.telegram.telegrambots.starter.SpringWebhookBot;
+
+public class HotDeskingBot extends SpringWebhookBot {
+    private final String BOT_USERNAME;
+    private final String BOT_TOKEN;
+    private final String BOT_PATH;
+    private final UpdateReceiver updateReceiver;
+
+    public HotDeskingBot(SetWebhook setWebhook, UpdateReceiver updateReceiver, String BOT_TOKEN, String BOT_USERNAME, String BOT_PATH) {
```
Comment (viktarkolbik) [inline src/main/java/com/exadel/coolDesking/bot/HotDeskingBot.java:14]:
```suggestion
    public HotDeskingBot(SetWebhook setWebhook, UpdateReceiver updateReceiver, String token, String username, String path) {
```
Don't use uppercase for method parameters

File: src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    private BotApiMethod<?> queryHandler(CallbackQuery query) {
+        log.info("DATA: {}", query.getData());
+        String telegramId = query.getFrom().getId().toString();
+        log.info("TelegramId: {}", telegramId);
+        User user = userService.getByTelegramId(telegramId);
+        log.info("UserState: {}", user.getUserState());
+        Handler handler = handlers.get(user.getUserState());
+        log.info("Handler: {}", handler);
```
Comment (viktarkolbik) [inline src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java:51]: your handler classes don't implement toString method. This log will print hashCode, it is senseless information.

Comment (gargatuma) [reply src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java:51]: line 51 will print out Handler class name, and that's what I want. I just want to see which handler was used for particular request)

File: src/main/java/com/exadel/coolDesking/bot/HotDeskingBot.java [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+    private final String BOT_TOKEN;
+    private final String BOT_PATH;
+    private final UpdateReceiver updateReceiver;
+
+    public HotDeskingBot(SetWebhook setWebhook, UpdateReceiver updateReceiver, String token, String username, String path) {
+        super(setWebhook);
+        this.updateReceiver = updateReceiver;
+        this.BOT_TOKEN = BOT_TOKEN;
+        this.BOT_USERNAME = BOT_USERNAME;
+        this.BOT_PATH = BOT_PATH;
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/HotDeskingBot.java:19]: you're not using params here :)

File: src/main/java/com/exadel/coolDesking/bot/config/botConfig.java [hunk 26]
```diff
@@ -0,0 +1,39 @@
...
+                .build();
+    }
+
+    @Bean
+    public HotDeskingBot HotDeskingSpringWebhookBot(
+            SetWebhook setWebhookInstance, UpdateReceiver updateReceiver,
+            @Value("${TELEGRAM_BOT_TOKEN}") String TELEGRAM_BOT_TOKEN,
+            @Value("${TELEGRAM_BOT_USERNAME}") String TELEGRAM_BOT_USERNAME,
+            @Value("${TELEGRAM_BOT_INTERNAL_URL}") String TELEGRAM_BOT_INTERNAL_URL,
+            @Value("${TELEGRAM_BOT_PATH}") String TELEGRAM_BOT_PATH) throws TelegramApiException {
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/config/botConfig.java:29]: use camelCase to name method params

File: src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+        BotResponseDto response= handler.handle(query, user);
+
+        userService.updateUserStatus(response.getUserState(), telegramId);
+        log.info(response.getMethod().toString());
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/coolDesking/bot/UpdateReceiver.java:55]: these should be `debug` - change later

**Review by ohalushchak-exadel — CHANGES_REQUESTED** [review]

**Review by ohalushchak-exadel — APPROVED** [review]

**Pull request merged**